### ISSUE TITLE
Research: upstream generation constraints to replace reactive guardrail accumulation

### ISSUE THREAD

**Issue by curdriceaurora** [curdriceaurora/Local-File-Organizer#850]
## Problem Statement

The current guardrail framework (465 CI tests, pre-commit hooks, governance meta-tests) is growing proportionally to generation inconsistency — not replacing it. The pattern is:

1. Generate code
2. Review finds anti-pattern (BLE violation, wrong PyPI version, vacuous assertion, missing newline)
3. Add guardrail to catch it next time
4. Generate code with a different anti-pattern
5. Repeat

Rules exist in `.claude/rules/feature-generation-patterns.md` and `.claude/rules/test-generation-patterns.md`. They are read during generation but treated as advisory — there is no structural barrier between planning and code emission. The pre-generation checklist is prose I can skip past.

**Root cause**: constraint is downstream of generation (review, CI) rather than upstream (before code is emitted).

## The Outlines Insight

`outlines` (dottxt-ai) constrains LLM token generation at inference time against a JSON schema or context-free grammar. A token that would violate the schema is structurally impossible — it is excluded from the sampling distribution. The constraint is upstream.

`outlines` does not work for this project as-is: it requires direct access to the model's sampling loop. Claude is an API black box — Anthropic does not expose logits or token distributions.

### Option A — Mandatory structured pre-generation contract (prompt protocol)

Before any `Write`/`Edit` tool call, Claude must emit a validated JSON contract:

```json
{
  "external_calls": [
    { "call": "embedder.transform(query)", "raises": "ValueError", "handler": "except ValueError" }
  ],
  "new_dependencies": [
    { "package": "rank-bm25", "version_floor": "0.2.0", "pypi_satisfiable": true }
  ],
  "test_assertions": [
    { "test": "test_top_k_limits_results", "top_k": 3, "assertion": "== 3", "vacuous": false }
  ],
  "exception_handlers": [
    { "handler": "except ValueError", "broad": false }
  ]
}
```

The contract is validated against the anti-pattern rules before code is written. Questions to answer:
- Can Claude Code hooks enforce tool-call ordering (ContractEmit must precede Write/Edit)?
- Can a `pre-tool` hook block `Write`/`Edit` unless a contract file exists in the session?
- What is the overhead per task?

### Option B — Two-agent adversarial pipeline

- Agent A generates code with no project-convention knowledge
- Agent B audits against rule files, outputs structured violation list
- Agents iterate until violations = 0
- User sees only the final output

Questions to answer:
- Does the Claude Code subagent model support this loop natively?
- What is the latency cost for a typical feature task?
- Can Agent B be given read-only access (no Write/Edit) to enforce the adversarial separation?

### Option C — Test-first strictly enforced

Tests are written and approved by the user before any implementation code is emitted. The test file IS the contract:
- Assertion types are explicit and visible before implementation exists
- Dependencies are declared before code imports them
- Exception handling scope is defined by what the test expects

Questions to answer:
- Can this be enforced via a Claude Code hook that checks for a corresponding test file before allowing Write to a `src/` path?
- How does this interact with TDD workflows where tests and implementation are iterated?

### Option D — Structured output via Claude tool_use schema

Use Claude's `tool_use` / `response_format` with a strict Pydantic schema for code generation tasks. While not true constrained generation (it's post-hoc validation with retry), it forces structured reasoning about each field before prose/code is emitted.

Questions to answer:
- Does filling out a strict schema before writing code materially reduce anti-pattern rate, even without token-level constraint?
- Can this be implemented as a Claude Code skill that wraps all implementation tasks?

### Option E — Claude Code hooks for tool-call ordering enforcement

Claude Code supports `pre-tool` and `post-tool` hooks. Evaluate:
- Can a `pre-tool` hook on `Write`/`Edit` check for a valid contract JSON in a session temp file?
- Can a hook reject the tool call (not just warn) if the contract is absent or invalid?
- What does the UX look like — does the hook block and prompt, or silently fail?

## Evaluation Criteria

For each option, assess:

| Criterion | Description |
|-----------|-------------|
| **Upstream-ness** | Does it constrain before or after generation? |
| **Enforceability** | Can it be made structurally impossible to bypass? |
| **Overhead** | Added latency/steps per feature task |
| **Inspectability** | Can the user review and reject the constraint artifact? |
| **Compatibility** | Works with Claude Code + Claude API (no local model required) |
| **Incremental adoptability** | Can it be added without rewriting existing workflow? |

## Expected Output

A recommendation for which option (or combination) to implement, with:
- Proof-of-concept implementation or hook configuration
- Measured anti-pattern rate before/after on a sample task
- Integration point in the existing PR workflow

## Context

- Anti-pattern rules: `.claude/rules/feature-generation-patterns.md`, `.claude/rules/test-generation-patterns.md`
- Guardrail layer: `tests/ci/` (465 tests), `.pre-commit-config.yaml`, `.claude/scripts/pre-commit-validation.sh`
- Triggered by: PR #846 generating 19 review findings across 3 rounds despite rules being in context
- Prior art: `outlines` (dottxt-ai), structured generation, constitutional AI, self-critique loops
